Do not sort operands as part of simplification [blocks: #3486]

tautschnig opened the pull request:
~Do-not-merge as above: this needs proper profiling to confirm the effect with data.~

kroening commented:
Does it make a difference whether or not the number of operands is large?
Also, is this perhaps about wrecking sharing? Or is the cost of the comparison too high for deep expressions?

tautschnig commented:
> Does it make a difference whether or not the number of operands is large?
Also, is this perhaps about wrecking sharing? Or is the cost of the comparison too high for deep expressions?

It's just the plain number of compare operations. 163M `sort_operands` calls, contributing to 1B `irept::operator<` calls, resulting in 4.5B `irept::compare` calls. I think close to zero of which are necessary. (Data for ReachSafety-BitVectors, the smallest of SV-COMP's categories, running for at most 600s per benchmark.)

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 51fcd8cbd957dafe2f2a1f1f00a73842e36048bd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106401199

tautschnig commented:
Combined with #1998 this increases the number of `symex_step` calls per second (on ReachSafety-ECA) from 2752.28 to 3749.87. Profiling them independently confuses gprof, because `sort_and_join` is invoked from different contexts.

smowton commented on src/util/simplify_utils.cpp:
```diff
@@ -119,30 +119,31 @@ static const struct saj_tablet &sort_and_join(
   return saj_table[i];
 }
 
-bool sort_and_join(exprt &expr)
```
! This needs a new name

tautschnig replied:
Why?

smowton replied:
Because it no longer necessarily sorts! A person reading `sort_and_join(x, false);` might reasonably suppose that regardless of what that `false` means, `x` is certainly now sorted.

smowton reviewed (approved):
This roughly halved a long-running case that simplifies a very large expression

smowton commented:
(but do still please rename `sort_and_join`)

martin-cs commented on regression/goto-analyzer/constant_propagation_07/test.desc:
```diff
@@ -3,7 +3,7 @@ main.c
 --constants --simplify out.gb
 ^EXIT=0$
 ^SIGNAL=0$
-^Simplified:  assert: 1, assume: 0, goto: 3, assigns: 13, function calls: 0$
```
This doesn't look right?  I can imagine it could alter the number of things simplified but not the number of assignments.

tautschnig replied:
Well, it just changed the number of assignments _that were simplified_ (well, "modified" perhaps would be a more accurate description as those "simplifications" were just re-arrangements of the operands).

martin-cs replied:
Ah, yes, good point.  Thanks.

martin-cs commented on src/util/expr.h:
```diff
@@ -9,17 +9,20 @@ Author: Daniel Kroening, kroening@kroening.com
+          it = as_const(expr).operands().begin(),                              \
+          it##_end = as_const(expr).operands().end();                          \
+        it != it##_end;                                                        \
+        ++it)
```
This seems to be an unrelated (but desireable and useful) change.

tautschnig replied:
Indeed, moved to #5553.

martin-cs reviewed (approved):
Seems reasonable.  Thanks @smowton for the performance comments.

tautschnig committed:
Do not sort operands as part of simplification

Data for ReachSafety-BitVectors, the smallest of SV-COMP's categories, running
for at most 600s per benchmark: 163M sort_operands calls, contributing to 1B
irept::operator< calls, resulting in 4.5B irept::compare calls. None of these
calls are actually necessary.

As part of this work, also clean up naming of "sort_and_join" functions
(as most of them don't actually sort or join, but instead just prepare
for doing so).

martin-cs commented on regression/cbmc-library/float-nan-check/test.desc:
```diff
@@ -6,20 +6,21 @@ main.c
-\[main.NaN.15\] line \d+ NaN on - in mynan - \(float\)n: SUCCESS
-\[main.NaN.16\] line \d+ NaN on \* in mynan \* \(float\)n: SUCCESS
-\[main.NaN.17\] line \d+ NaN on / in mynan / \(float\)n: SUCCESS
+\[main.NaN.6\] line \d+ NaN on \+ in myinf \+ -myinf: FAILURE
```
Actually; this raises an interesting question.  Floating-point addition and multiplication are commutative but not associative.  If seems like normalising / sorting these would be worth while.

tautschnig replied:
I'm not sure I am able to follow: why would sorting (which this PR largely disables) be particularly beneficial for those operations? We surely can selectively re-enable sorting, but I don't think I understand the rationale just yet.

martin-cs replied:
```
float a;
float b;
if (!isnan(a) && !isnan(b)) {
  assert(a + b === b + a);
}
```

Normalising the order of commutative-but-not-associative operations means we can resolve this at simplification, rather than an expensive solver call.  It is not a massively common case but it is very cheap to implement and IIRC is a huge speed up in this case.  Note that this can be purely local as it is not associative.

tautschnig replied:
It seems we didn't actually simplify such expressions before as commutative-but-not-associative operations were exempt. I have now added 375f208 to take care of this, thank you very much for raising this!

martin-cs reviewed (approved):
Yes but handling 'commutative but not associative' would be good.

tautschnig committed:
jbmc-strings/ConstantEvaluationTrim: avoid null dereference

String refinement may or may not be successful (resulting in "ERROR"
results rather than "FAILURE") when working on an unconstrained string
(which the null dereference yields), depending on the models returned by
the SAT solver.

tautschnig committed:
Simplify: sort operands for commutative, but not associative operations

Testing floating-point equality is expensive, and thus warrants handling
special cases (even if they might be rare) in the simplifier. This
(simplistic) test would take 5 seconds with CaDiCaL as SAT back-end when
not performing simplification.
